### PR TITLE
Fix a bug related to precision in inline_linear and other clean up

### PR DESCRIPTION
**Pull request by cdcapano**
This patch fixes a bug in `fd_decompress` that would cause it to use the wrong precision if the compressed waveform was saved as double precision, but the requested output is single precision. Currently, what code to use for interpolation is being determined by the precision of the sample frequencies. However, it should be determined by the precision of the output vector. The result is that if sample frequencies are double precision but the output is single, the double precision interpolation code gets called. The result is the output pointer is of the wrong type, and so the results are written to the wrong parts in memory. The fix is a one-liner at 523.

This patch also speeds up compress_waveform by removing needless frequency vector generation. It also ensures that added points are new (this probably wasn't a problem before, but this ensures it). Finally, this patch also fixes a minor bug in the `inline_linear` interpolation code: the break when `next_sfindex == hlen` should be a break out of the for loop, not the second while loop. I don't think that had any effect on anything, but worthwhile to fix now.

This patch relies on #1473, so putting on hold until that patch is merged.

### CODE REVIEW

**Comment by cdcapano** [comment]
Ok, this is ready to push now. Since the location of the break statement was fixed in #1473, this now just fixes the precision bug and speeds up the compression algorithm.

File: pycbc/waveform/compress.py [hunk 1]
```diff
@@ -257,7 +257,16 @@ def compress_waveform(htilde, sample_points, tolerance, interpolation,
         minpt = vecdiffs.argmax()
         # add a point at the frequency halfway between minpt and minpt+1
         add_freq = sample_points[[minpt, minpt+1]].mean()
-        addidx = int(add_freq/df)
+        addidx = int(round(add_freq/df))
+        # ensure that only new points are added
+        while addidx in sample_index:
+            addpt -= 1
```
Comment (soumide1102) [inline pycbc/waveform/compress.py]: @cdcapano Landscape is complaining about addpt not being defined before. Should it be set to a value -1 somewhere before the while loop?

File: pycbc/waveform/compress.py [hunk 2]
```diff
@@ -257,7 +257,16 @@ def compress_waveform(htilde, sample_points, tolerance, interpolation,
...
+            try:
+                minpt = diffidx[addpt]
```
Comment (soumide1102) [inline pycbc/waveform/compress.py]: Should we set `diffidx=vecdiff.argsort()` before the while loop?

**Comment by cdcapano** [comment]
@soumide1102 Ok, I think I fixed the issues.

**Pull request merged**